Big code refactor

joao-vicente opened the pull request:
I started by making a small improvement (to create a small PR) to the `@qualweb/locale` package and ended up overly excited. One thing led to another and here we are with a PR with more than 300 files... sorry not sorry 😅
In the end, a lot of ideas that I had previously were implemented, and even tho this is a big change, I'm far from done. The goal is to improve future developments. A lot of testing is still needed but I'm opening already a draft PR so if anyone has time, to start questioning it!

Major points of this refactor:
- Code declutter (removed a lot of useless code, and files);
- Better abstractions (especially on the evaluation modules);
- Better typing;
- Better separation of responsibilities of each module/class.

Tellus commented:
Hey @joao-vicente!

Apologies for the silence.

I hope to have some time to go through the commit tomorrow or Thursday. As We discussed, I'm going to check out the changes in a copy of the main branch and see if I can't organize them in more granular commits.

Tellus commented:
Alright, so I've been trying to get a grasp of the many changes in the commit. I think I've managed to narrow down the changes into a bunch of better fitting commits. To my mind, each point could correspond to roughly one commit. I'd argue that commits shouldn't necessarily be as tiny/atomic as possible. My preferred approach is to keep each commit to one "point" or "thing" that can be concisely described in the first line of the commit, so other developers can see - at a glance - what changes have happened. When working locally with rebasing, it can also save a ton of merge headaches since git can interweave the changes with your own instead of forcing one big change into your history.

- chore: update contributor email in package.json
- chore: update copyright year in license
- refactor: prefer ESM-style exports over CommonJS
- refactor: don't implement a constructor that solely calls its base class' constructor
- refactor: prefer using `this` over `super` when calling methods (less debugging if one of those methods should get overridden in future)
- refactor: avoid `typeof` typing when direct/original type is available
- feat: remove `ACTRuleDecorator` - its logic should be part of a constructor, not a decorator
- refactor: move classes out of index.ts files into their own separate files
- feat: deprecate types package *(this should probably be done over several refactor commits before removing the package entirely)*
- feat: new package, `lib`
- feat: `locale` package API changed. BREAKING CHANGE!
- chore: migrate webpack configurations to TypeScript

There are bigger changes that I don't know how to document or narrow down from the single commit.

For example, when we talked, you described how you merged several packages into others because it seemed unnecessary to have a whole package for them (evaluation and dom?). I'm not sure I can pull out those changes from the single commit because I can't see specifically what that exact change encompasses when mixed with the others.

Finally, I'm not sure it's a good idea to commit developer preferences to the workspace file. Right when I opened your changeset my VSCode went from a calm dark to bright whites - had me blinded for a good few moments :D

One thing I should try and do today is upload a CONTRIBUTING.md file with instructions on writing changesets. I thought I'd already added a basic/boilerplate explanation in the README file but that's clearly not the case.

joao-vicente commented:
Amazing @Tellus! I think I wouldn't be able to create a summary that concise 😅 
I'm glad you managed to find the time to do it, and it is easier to tackle everything separately now. I didn't/don't have much time these two weeks (the previous one, and this one) so I don't think I'll be able to contribute much... On the next one I expect to have more free time to help you with this, so just let me know how you'd like to proceed.

Regarding the VSCode preferences... My bad 😅 I didn't understand that I included those in the commit.. but I'm still a bit confused because I also don't use bright whites. I use the One Dark Pro theme 😉

Tellus commented:
Hey, João

I'm sorry for the silence. Easter came around - and then a vicious ear infection.

I'm having some trouble with redoing the one commit as the separate I suggested. The problem, I think, boils down to how Git "hunks" for files work. I was hoping that I could check out all the changes from your commit as uncommitted changes in a new branch, and then selectively just commit specific lines of changes in batches.

This works for some of the changes, but when there are several overlapping changes that should belong in different commits (for example, changes to imports at the top of a file and the immediate export of a class on the next line) can't be cleanly separated with quick Git actions. There's an interactive patch command in Git that lets you do this, but going through 304 files line by line - for each commit - is just too much work :P I couldn't find a non-interactive version that could be easily batched to do the same.

So I think we're kind of left with either merging everything as a single commit (I'm still not a fan, but as you said, you ran all relevant unit tests to ensure that code should work as expected) or re-doing all the work by hand, separated by change type. I.e. make a new branch from main, change all export statements, commit, change all type usages, commit, clean up imports, commit - and so on.

Assuming the first, we should probably still address the changes that definitely should not be committed (deleted CHANGELOG files, personal user settings in the workspace file, etc). Do you think you could amend/adjust those or should we try to speed-run the individual changes, you think?

Tellus commented:
Hey João.

Just quickly noting down what we talked about (mostly because my memory's bad :D).

We'll just pull everything in as a single PR, but with adjustments for how CI/CD works. Accounting for how the merge conflicts look right now, I think this is what we arrived at:

- Create a new branch from `main`, then copy over all changes from this PR's commit, so they pop up as uncommitted changes.
- *Undo* your changes to package.json files and CHANGELOG files. These portions are updated automatically in a Github action
- *Undo* the changes to the workspace file (the user-specific additions)
- *Undo* the changes to the lockfile, to avoid any conflicts. *Do* rebuild it with `npm install` in a separate commit afterwards (I think this should be fine).
- *Do* change your collaborator entry in the relevant package.json files. Attribution is important :)
- Since you're just checking out changes from another commit, you may have to go over the source files you intended delete again. **Since we're effectively stopping the updates for some public packages, we should consider making a release that mentions their deprecation in their README files and NPM.js descriptions before a commit that fully deletes them.**

Once that's done, we'll need to adjust the CI/CD pipeline to the added/deleted packages. Let's cross that bridge once the other changes are done. 

Thanks for the chat and have a great weekend!

joao-vicente commented:
Hey @Tellus, it took me a while, but I've finished all the changes!

I had to go over a lot more files than expected because the packages were not compiling correctly. The package `@qualweb/types` was the culprit here because it started interfering with the new changes to the typing system. In the end, I had to get rid of it. So I'll try to summarise all the changes I made:
- I've defined an order in the global `package.json` in which the packages should be compiled;
- The new package I had created `@qualweb/lib`, I undid it and created a shared folder (`shared`) with types/classes/methods to be shared. This is not a package;
	- Depending on how it goes, the types may be moved to their respective module/package in the future;
-   The `@qualweb/types` is no longer needed;
- Updated all other packages to stop using `@qualweb/types` and use the new `shared` folder;
- Updated the `tsconfig.json` files to use path aliases to reference the new `shared` folder and data from other packages;
- Updated the `webpack.config.ts` files to take into account the path aliases during the compilation;
- Added the `Window` types to all packages that needed it, but each one with only the variables they use.
- Some TypeScript issues fixed;
- Create a new class `Guideline` so that in the future we abstract the rules/techniques/best-practices;
	- Still not sure about that name (naming is hard. Open to suggestions);
	- I didn't want to change even more files, so I simply created the new class but didn't move any code. We're starting to have too many abstractions and that's generally not good/needed (e.g. Guideline -> Rule -> AtomicRule -> QW_ACT_R1). To be improved in the future.

What I didn't do:
- Version bumps;
- Package published.

I was thinking in publishing new versions with a minor version increase and appending `-alpha`. This PR has a lot of changes And it would be better to test it "officially" before merging into the main branch.

Let me know if you have some time to do an overall review of these latest changes, and if you notice any red flags in this new approach, of course, let me know what you think ;)

Have a nice week!

Tellus commented:
Sorry for the silence! I was expecting Github to ping me via e-mail when you @mentioned me - nope!

I'll take a speedrun through it this evening. It's probably fine :-)

Tellus commented:
I just double checked my points from way back when.

- changelogs/package.json reverted. Excellent.
- the workspace file is nice and clean. My eyes are thankful :)

Pop a changeset in there to document the changes for the CI/CD to handle, and I'm all set.

Unless @carlosapaduarte has any objections I say we just merge everything and deal with the fallout if it happens :P

We should probably still think about how to signal the discontinuation of the removed packages, just so any direct consumers of them gets notice next time they try to update or run `npm i` in their own projects.

Tellus commented:
I've merged in the changes from the second PR.

Depending on what I missed, these would be the last TODOs. Some can probably wait until after a merge.

- Finish up the changeset/changelog. NOTABLY document the changes to using @qualweb/core
- Fix merge conflicts: NOTABLY the hot fixes made by Carlos since we started this refactor business.
- @qualweb/act-rules: validation tests should be checked. Are the errors new or is the main version also failing those tests?
- @qualweb/cli: unit tests, JSON file with options as input.
- eslint config needs touching up. It runs now, which is good, but the sonarjs plugin (maybe others) insist on parsing the final files in "dist" - we're only interested in linting the raw source files.

Tellus commented:
Resolved merge conflicts between branch `main` and the PR branch `qw-locale-improvements`.

**There are still a few things to iron out.** API calls that haven't been updated in the interim commits, unit tests.

On it.

Tellus commented:
Only one thing left, I think. The module unit test in wcag-techniques (module.spec.ts) is failing, and I couldn't quickly figure out why. Looking into it.

I've updated this PR to merge into `develop` instead, so we can safely merge when happy and see how it performs via NPM.

Tellus commented:
I've fixed the tests in @qualweb/wcag-techniques (and a few other places) by exporting the `WcagTechniquesModule` in a separate subpath.

There's a lot of code that will try to use the `window` global when it's simply imported, even it's not going to be run in the browser. This causes run-time errors much higher up and is hard to diagnose.

After merging this PR it might be prudent to look into reducing use of the `window` global where possible. In a lot of cases, it's a use of `AccessibilityUtils` and `DomUtils`, which can just be imported directly instead of relying on a (possibly undefined) global variable.
